prow/plugins/cla/cla_test.go

```diff
@@ -161,7 +190,10 @@ func TestCLALabels(t *testing.T) {
 			SHA:     tc.statusSHA,
 			State:   tc.state,
 		}
-		if err := handle(fc, logrus.WithField("plugin", pluginName), se); err != nil {
+		cc := plugins.CLAConfig{
+			CLAContextNames: []string{"cla/linuxfoundation", "EasyCLA"},
+		}
+		if err := handle(fc, logrus.WithField("plugin", pluginName), se, cc); err != nil {
```

Comment:
OK, I think I assumed there was less duplication of logic before

So the way I read the StatusEvent handling code, is that it will respond to whatever the status is.  So if the CLA plugin is configured to listen to two CLA statuses, we will have a race between which of the two statuses reports in first.  Slowest status wins.

I think what would provide more consistency is to try to de-dupe the logic and do what `handleComment` does: get the PR's combined statuses, and ensure one of them consistently wins the race.  

e.g. let's say cla/linuxfoundation consistently wins:
```yaml
# happy path
state: 
  cla/linuxfoundation: pending
  EasyCLA: pending
  labels: []
receive:
  cla/linuxfoundation: success
result:
  add: "cncf-cla: yes"

# conflict: EasyCLA arrives last
state: 
  cla/linuxfoundation: success
  EasyCLA: pending
  labels: ["cncf-cla: yes"]
receive:
  EasyCLA: failure
result:
  # do nothing

# conflict: cla/linuxfoundation arrives last
state: 
  cla/linuxfoundation: pending
  EasyCLA: success
  labels: ["cncf-cla: yes"]
receive:
  cla/linuxfoundation: failure
result:
  add: "cncf-cla: no"
  remove: "cncf-cla: yes"
```

Reply:
>  So if the CLA plugin is configured to listen to two CLA statuses

From the docs, `The most recent status for each context is returned`.
I think my question is - are `EasyCLA` and `cla/linuxfoundation` considered as two different contexts or different names for a single `CLA` context? How does getting combined status here work? Combined status is still going to return both the contexts - how do we know which one is the latest?

Reply:
>I think my question is - are EasyCLA and cla/linuxfoundation considered as two different contexts or different names for a single CLA context

Well, I'm assuming I'm going to see two entries, one called `cla/linuxfoundation` and one called `EasyCLA` while the CNCF has both things running simultaneously on their end.  So they're different contexts.  You're updating the plugin to decide which one to treat as the authoritative source of truth from the CNCF.  This allows us to have the plugin pay attention to the new EasyCLA check, but defer to the old cla/linuxfoundation check as authoritative, without having to change anything.

Another possibly simpler option is to roll back some of what you've done, and update the plugin to listen to one-and-only-one context.  In this scenario we would:
- make the plugin listen to a configurable context
- start with cla/linuxfoundation as this plugin's source of truth
- setup EasyCLA on the CNCF's end to start reporting to PRs
- setup prow/branch-protector to treat EasyCLA contexts as non-blocking
- wait
- generate a report that shows over time, for all PRs within a given window, PRs where cla/linuxfoundation and EasyCLA differed
- figure out why and fix
- wait again
- generate report again
- repeat until there is no difference, and then we change branch-protector to make the EasyCLA context required, the cla/linuxfoundation context non-blocking, and the cla plugin to listen to EasyCLA to apply "cncf-cla: yes" labels

Reply:
>How does getting combined status here work? Combined status is still going to return both the contexts - how do we know which one is the latest?

What I was saying above is that the `handle` function _doesn't call GetCombinedStatus_, it only consider the status from the StatusEvent it's handling.  So it will get called once whenever `cla/linuxfoundation` is update, and again whenever `EasyCLA` is updated, and there's no way to guarantee which of those is going to arrive first.

Calling GetCombinedStatus to get both contexts is the point.  We don't _care_ which one came in latest, we only care that it came in for the commit we're considering, and that the authoritative context (if set) always overrides the others.

Again, I'm thinking this may all be too confusing, and it might be simpler to support a single (configurable) context only.  But we need to see how EasyCLA behaves live (in terms of responsiveness, reliability, consistency, and agreement with cla/linuxfoundation) before I will ever consider changing our workflow to pay attention to that over cla/linuxfoundation.

Reply:
EasyCLA has been enabled for the [k/org](https://github.com/kubernetes/org) repo in a non-blocking way as a test.
For some additional information, EasyCLA is currently in use by 3 other CNCF projects:
- Open telemetry (47 repos)
- grpc (26 repos)
- cloud custodian (6 repos)

For gRPC it's been some time (most of the kinks in the project were worked out with this one), I don't know when it was enabled for Otel or cloud custodian.

The LF has said if we do have problems we can roll back.